aten/src/ATen/Exceptions.h

```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#include <ATen/ATen.h>
+
+#include <typeinfo>
+
+
+namespace at {
+namespace checks {
+
+inline bool tensor_on_same_device(const at::Tensor &ten1, const at::Tensor &ten2){
+  return ten1.get_device()==ten2.get_device();
+}
+
+inline bool tensor_on_same_device(
+  const at::Tensor &ten1,
+  const c10::optional<at::Tensor> &ten2
+){
+  return !ten2.has_value() || ten1.get_device()==ten2->get_device();
+}
+
+inline bool tensor_on_cpu(const at::Tensor &ten){
+  return !ten.is_cuda(); // TODO: Should be a better way to do this
+}
+
+inline bool tensor_on_cpu(const c10::optional<at::Tensor> &ten){
+  return !ten.has_value() || !ten->is_cuda(); // TODO: Should be a better way to do this
+}
+
+inline bool tensor_on_cuda_gpu(const at::Tensor &ten){
+  return ten.is_cuda();
+}
+
+inline bool tensor_on_cuda_gpu(const c10::optional<at::Tensor> &ten){
+  return !ten.has_value() || ten->is_cuda();
+}
+
+}}
+
+#define TENSORS_HAVE_SAME_NUMEL(x,y) \
```

Comment:
If these macros are going to be in a public header, they can't be given a generic prefix like `TENSORS_`; will have to be `TORCH_` something. Probably `TORCH_CHECK_...`